### PR TITLE
#18 Add flags to configure timeouts

### PR DESCRIPTION
**Pull request by klebervirgilio**
Issue #18 

Hi,
You will find a few `TODO: TBC` reminding us to define the timeouts.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,7 +37,12 @@ The following flags are available for configuring PlanB on start-up:
   be sent to stdout. Default value is ``./access.log``.
 - ``--request-timeout``: Total backend request timeout in seconds. Default
   value is ``30``.
-- ``--dial-timeout``: Dial backend request timeout in seconds. Default value is
```
Comment (cezarsa) [inline README.md]: Was this line removed by mistake?

File: main.go [hunk 2]
```diff
@@ -264,6 +264,26 @@ The value 'none' can be used to disable access logs.`),
 			Usage: "Dial backend request timeout in seconds",
 		},
 		cli.IntFlag{
+			Name:  "client-read-timeout",
+			Value: 30, // TODO: TBC
```
Comment (cezarsa) [inline main.go]: I'd prefer to have the default value for these new flag set to `0`. This way we don't alter the default behavior for planb when starting a new instance without setting them.

**Comment by andrestc** [comment]
@klebervirgilio I've droppped go 1.6 and 1.7 from our travis build, and added go 1.8 since it's now required. Can you rebase your PR from master ?

File: main.go (hunk 2 above)
Comment (klebervirgilio) [inline main.go]: Good point. 
I'm happy to change. Should I?

File: main.go [hunk 3]
```diff
@@ -264,6 +264,26 @@ The value 'none' can be used to disable access logs.`),
...
+			Usage: "Maximum duration for reading the entire request, including the body",
+		},
+		cli.IntFlag{
+			Name:  "client-read-header-timeout",
+			Value: 30, // TODO: TBC
```
Comment (klebervirgilio) [inline main.go]: @cezarsa @andrewsmedina I left a few `TODO: TBC` so we can discuss the default values for the new timeouts.

**Comment by andrestc** [comment]
I think we can keep 0 as the default value for these timeouts as @cezarsa pointed out, then i think its good to go.

**Comment by klebervirgilio** [comment]
Cool. Timeouts set to 0.

**Comment by cezarsa** [comment]
You'll need to read the flag values and set them in the `reverseproxy.ReverseProxyConfig` instance. Similar to the DialTimeout in https://github.com/tsuru/planb/blob/master/main.go#L118.

Also, I think it might be better to either change from `cli.IntFlag` to `cli.DurationFlag` for these new timeout flags or change the description to make it explicit that the unit is seconds and multiply the value appropriately. (i.e. "Maximum duration for reading the entire request, including the body, **in seconds**")

**Comment by andrestc** [comment]
I think we should use cli.DurationFlag since one might want to set those timeouts for sub-seconds time.

**Comment by klebervirgilio** [comment]
Agreed. I changed the flags to `cli.DurationFlag`. Plus, I set them in the `reverseproxy.ReverseProxyConfig`.

File: main.go [hunk 4]
```diff
@@ -112,11 +112,15 @@ func runServer(c *cli.Context) {
...
-		RequestIDHeader: c.String("request-id-header"),
-		FlushInterval:   time.Duration(c.Int("flush-interval")) * time.Millisecond,
-		DialTimeout:     time.Duration(c.Int("dial-timeout")) * time.Second,
-		RequestTimeout:  time.Duration(c.Int("request-timeout")) * time.Second,
+		Router:            &r,
+		RequestIDHeader:   c.String("request-id-header"),
+		FlushInterval:     time.Duration(c.Int("flush-interval")) * time.Millisecond,
+		DialTimeout:       time.Duration(c.Int("dial-timeout")) * time.Second,
+		RequestTimeout:    time.Duration(c.Int("request-timeout")) * time.Second,
+		ReadTimeout:       c.Duration("client-read-timeout") * time.Second,
```
Comment (andrestc) [inline main.go]: You dont need to do `* time.Second` since the flag is already a duration, you need to pass it directly.

Comment (klebervirgilio) [inline main.go]: Good point! Fixed.

File: README.md [hunk 5]
```diff
@@ -37,8 +37,14 @@ The following flags are available for configuring PlanB on start-up:
...
-  ``10``.
+- ``--dial-timeout``: Dial backend request timeout in seconds. Default value is ``10``.
+- ``--client-read-timeout``: Maximum duration for reading the entire request, including the body. Default
+  value is ``0``.
+- ``--client-read-header-timeout``: Amount of time allowed to read request headers. Default value is ``0``.
+- ``--client-writer-timeout``: Maximum duration before timing out writes of the response. Default
```
Comment (andrestc) [inline README.md]: I think this should be `--client-write-timeout` instead of `writer`

File: main.go [hunk 6]
```diff
@@ -112,11 +112,15 @@ func runServer(c *cli.Context) {
...
-		DialTimeout:     time.Duration(c.Int("dial-timeout")) * time.Second,
-		RequestTimeout:  time.Duration(c.Int("request-timeout")) * time.Second,
+		Router:            &r,
+		RequestIDHeader:   c.String("request-id-header"),
+		FlushInterval:     time.Duration(c.Int("flush-interval")) * time.Millisecond,
+		DialTimeout:       time.Duration(c.Int("dial-timeout")) * time.Second,
+		RequestTimeout:    time.Duration(c.Int("request-timeout")) * time.Second,
+		ReadTimeout:       c.Duration("client-read-timeout"),
+		ReadHeaderTimeout: c.Duration("client-read-header-timeout"),
+		WriteTimeout:      c.Duration("client-writer-timeout"),
```
Comment (andrestc) [inline main.go]: I think this should be `client-write-timeout` instead of `writer`

**Comment by andrestc** [comment]
looking good, except for two nits

**Comment by klebervirgilio** [comment]
Can't believe I missed that. Typos fixed. Thanks.

**Comment by cezarsa** [comment]
Thank you @klebervirgilio !

**Pull request merged**